specs/core/0_beacon-chain.md

```diff
@@ -547,385 +534,469 @@ class BeaconState(Container):
+    end = (len(indices) * (index + 1)) // count
+    return [indices[shuffle_index(ValidatorIndex(i), len(indices), seed)] for i in range(start, end)]
 ```
 
-### `is_slashable_validator`
+#### `validate_indexed_attestation`
 
 ```python
-def is_slashable_validator(validator: Validator, epoch: Epoch) -> bool:
+def validate_indexed_attestation(state: BeaconState, indexed_attestation: IndexedAttestation) -> None:
     """
-    Check if ``validator`` is slashable.
+    Verify validity of ``indexed_attestation``.
     """
-    return (not validator.slashed) and (validator.activation_epoch <= epoch < validator.withdrawable_epoch)
+    bit_0_indices = indexed_attestation.custody_bit_0_indices
+    bit_1_indices = indexed_attestation.custody_bit_1_indices
+
+    # Verify no index has custody bit equal to 1 [to be removed in phase 1]
+    assert len(bit_1_indices) == 0
+    # Verify max number of indices
+    assert len(bit_0_indices) + len(bit_1_indices) <= MAX_VALIDATORS_PER_COMMITTEE
+    # Verify index sets are disjoint
+    assert len(set(bit_0_indices).intersection(bit_1_indices)) == 0
+    # Verify indices are sorted
+    assert bit_0_indices == sorted(bit_0_indices) and bit_1_indices == sorted(bit_1_indices)
+    # Verify aggregate signature
+    assert bls_verify_multiple(
+        pubkeys=[
+            bls_aggregate_pubkeys([state.validators[i].pubkey for i in bit_0_indices]),
+            bls_aggregate_pubkeys([state.validators[i].pubkey for i in bit_1_indices]),
+        ],
+        message_hashes=[
+            hash_tree_root(AttestationDataAndCustodyBit(data=indexed_attestation.data, custody_bit=0b0)),
+            hash_tree_root(AttestationDataAndCustodyBit(data=indexed_attestation.data, custody_bit=0b1)),
+        ],
+        signature=indexed_attestation.signature,
+        domain=get_domain(state, DOMAIN_ATTESTATION, indexed_attestation.data.target.epoch),
+    )
 ```
 
-### `get_active_validator_indices`
+#### `slot_to_epoch`
 
 ```python
-def get_active_validator_indices(state: BeaconState, epoch: Epoch) -> Sequence[ValidatorIndex]:
+def slot_to_epoch(slot: Slot) -> Epoch:
     """
-    Get active validator indices at ``epoch``.
+    Return the epoch number of ``slot``.
     """
-    return [ValidatorIndex(i) for i, v in enumerate(state.validators) if is_active_validator(v, epoch)]
+    return Epoch(slot // SLOTS_PER_EPOCH)
 ```
 
-### `increase_balance`
+#### `epoch_first_slot`
 
 ```python
-def increase_balance(state: BeaconState, index: ValidatorIndex, delta: Gwei) -> None:
+def epoch_first_slot(epoch: Epoch) -> Slot:
     """
-    Increase validator balance by ``delta``.
+    Return the first slot of ``epoch``.
     """
-    state.balances[index] += delta
+    return Slot(epoch * SLOTS_PER_EPOCH)
 ```
 
-### `decrease_balance`
+#### `delayed_activation_exit_epoch`
 
 ```python
-def decrease_balance(state: BeaconState, index: ValidatorIndex, delta: Gwei) -> None:
+def delayed_activation_exit_epoch(epoch: Epoch) -> Epoch:
```

Comment:
This sounds like the property of an object to me as it is a common noun. I think `get_delayed_activation_exit_epoch` is better.

Reply:
Agreed it's more descriptive locally. I'm trying to reserve the `get_` prefix for state accessors.

Reply:
I thought as much and I really like that idea. I still think we should find an alternate prefix to use within the `Misc` section. Ideas I've had thus far are:
* Compute
* Calculate
* Determine
* Obtain
* Acquire
* Compute

Reply:
`compute_activation_exit_epoch` ?

Reply:
Agreed that "delayed" doesn't add much value here

Reply:
I like that! I would like to change all of them in the `Misc` section, but let's merge this first.